### PR TITLE
[LIP-3] incorporate feedbacks

### PR DESCRIPTION
**Pull request by zekun000**

### CODE REVIEW

File: lips/lip-3.md [hunk 1]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
 ---
 
 # Introduction
-Libra Upgrade Management (LUM) is a set of protocols for the Libra Networks (LN) company to reconfigure the blockchain. With respect to time-sensitive decisions and incident response, LUM provides the Libra Networks--a subsidiary of the Libra Association licensed in Switzerland to operate a payment system--with technical means to respond to critical incidents, as (i) necessitated by its license terms and the law and (ii) as determined by the Libra Network risk appetite/tolerance policy.
+Libra Upgrade Management (LUM) is a set of protocols to reconfigure the blockchain. With respect to time-sensitive decisions and incident response, LUM provides the Libra Networks--a subsidiary of the Libra Association licensed in Switzerland to operate a payment system--with technical means to respond to critical incidents, as necessitated by compliance.
```
Comment (aching) [inline lips/lip-3.md:11]: Does the -- render nicely and intentional?

File: lips/lip-3.md [hunk 2]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
 
 Due to being a reserve backed currency, the true blockchain is the one that is maintained by the Libra Association and reflects the assets in the Libra Reserve. In other words, in contrast to Bitcoin and other “proof of work” and “proof of stake” networks where the miners and users beliefs determine the current fork of the blockchain, with Libra the blockchain used by the Libra Reserve determines the true blockchain. Furthermore, Libra Networks operates as a regulated payment company under a Swiss license and must be able to enforce its policy and oversight, as well as to respond to critical operational incidents.
 
-The goal of LUM is to provide the Association with means to exercise its authority to reconfigure the blockchain and to effectively remediate any plausible operational incident while maintaining the trust gained through decentralized governance and distributed systems.
+The goal of LUM is to provide the means to effectively remediate any plausible operational incident in a transparent manner.
```
Comment (aching) [inline lips/lip-3.md:17]: LUM is both for normal upgrades as well as quickly resolve incidents right?

```suggestion
The goal of LUM is to provide the means to effectively and transparently update the software in order to meet changing requirements as well as remediate any plausible operational incident.
```

File: lips/lip-3.md [hunk 3]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
 We first present all building blocks and the reasoning, followed by concrete examples.
 
 # Building blocks
-There’s two types of upgrades in the system - (i) updates to the Libra protocol software and (ii) changes to on-chain state. We discuss them correspondingly.
+There are two types of upgrades in the system - (i) updates to the Libra protocol software and (ii) changes to on-chain state. We discuss them correspondingly.
 
-## Software upgrade
+All building blocks may be used for "business-as-usual" upgrades or as urgent measures.
```
Comment (aching) [inline lips/lip-3.md:23]:
```suggestion
All building blocks may be used for "business-as-usual" upgrades or to address urgent measures.
```

File: lips/lip-3.md [hunk 4]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
+
+## (i) Software upgrade
 Specification changes go through the LIP process and non-specification changes (e.g. bug fixes and performance improvements) do not need to initiate the LIP process. All software changes are submitted through standard Github reviews and proceed through the code review process before being merged to master.
 
-Periodically, LN needs to release a new version of software and coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per instructions.
+To release a new version of software LN needs to coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per instructions.
```
Comment (aching) [inline lips/lip-3.md:28]:
```suggestion
In order to release a new version of software, LN needs to coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per the provided instructions.
```

File: lips/lip-3.md [hunk 5]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
 
-Some specification changes may require all validators to transition at the same height, which is accompanied with an on-chain state value update (described in the next section).
+Some changes may require all validators to transition at the same height, which is accompanied with an on-chain state update (described in (ii)).
 
-## On-chain state upgrade
-On-chain state can only be updated via transactions. We propose three ways for LN to execute critical updates, ordered by most frequently expected to least frequently expected: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
+## (ii) On-chain state upgrade
+On-chain state can only be updated via transactions. We propose three ways to execute critical updates, ordered by most frequently to least frequently: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
```
Comment (aching) [inline lips/lip-3.md:33]:
```suggestion
On-chain state can only be updated via transactions. We propose three ways to execute critical updates, ordered by most frequent to least frequent: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
```

File: lips/lip-3.md [hunk 6]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
 
 ### Normal multi-sig transactions
-Normal transactions go through VM and smart contracts, both the resulting state change and the authenticating set of signatures are transparent and visible on-chain for auditing.
-
-Different groups would authenticate different types of transactions. For example, there would be a group for compliance transactions and another group for validator set management.
+Normal transactions go through VM and invoke Libra Framework modules (see LIP-2) to produce a Writeset that alters ledger state. Both the resulting state change and the authenticating set of signatures are transparent and visible on-chain for auditing.
```
Comment (aching) [inline lips/lip-3.md:36]: Maybe add the link to LIP-2?

```suggestion
Normal transactions go through the VM and invoke Libra Framework modules (see LIP-2) to produce a Writeset that alters ledger state. Both the resulting state change and the authenticating set of signatures are transparent and visible on-chain for auditing.
```

File: lips/lip-3.md [hunk 7]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
 
-The goal of LUM is to provide the Association with means to exercise its authority to reconfigure the blockchain and to effectively remediate any plausible operational incident while maintaining the trust gained through decentralized governance and distributed systems.
+The goal of LUM is to provide the means to effectively remediate any plausible operational incident in a transparent manner.
 We first present all building blocks and the reasoning, followed by concrete examples.
 
 # Building blocks
-There’s two types of upgrades in the system - (i) updates to the Libra protocol software and (ii) changes to on-chain state. We discuss them correspondingly.
+There are two types of upgrades in the system - (i) updates to the Libra protocol software and (ii) changes to on-chain state. We discuss them correspondingly.
```
Comment (aching) [inline lips/lip-3.md:21]: Maybe consider making i and ii links to the other sections?

File: lips/lip-3.md [hunk 8]
```diff
@@ -8,52 +8,52 @@ created: 06/09/2020
...
+Libra Upgrade Management (LUM) is a set of protocols to reconfigure the blockchain. With respect to time-sensitive decisions and incident response, LUM provides the Libra Networks--a subsidiary of the Libra Association licensed in Switzerland to operate a payment system--with technical means to respond to critical incidents, as necessitated by compliance.
 
-It should be noted that the Association appoints LN officers. Officers are accountable for their actions, and usual measures may be taken upon deviation from their expected conduct.  Hence, the ultimate responsibility and power is by distributed governance of the Association.
+It should be noted that the Association appoints Libra Network(LN) officers. Officers are accountable for their actions, and usual measures may be taken upon deviation from their expected conduct.  Hence, the ultimate responsibility and power is by distributed governance of the Association.
```
Comment (aching) [inline lips/lip-3.md:13]: I personally feel like this sentence is not important for this LIP and can be removed.

**Review by aching — CHANGES_REQUESTED** [review]
Looks pretty good, a few nits.

File: lips/lip-3.md [hunk 9]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
+There are two types of upgrades in the system - (i) [updates to the Libra protocol software](#software-upgrade) and (ii) [changes to on-chain state](#on-chain-state-upgrade). We discuss them correspondingly.
+
+All building blocks may be used for "business-as-usual" upgrades or to address urgent incidents.
 
 ## Software upgrade
 Specification changes go through the LIP process and non-specification changes (e.g. bug fixes and performance improvements) do not need to initiate the LIP process. All software changes are submitted through standard Github reviews and proceed through the code review process before being merged to master.
 
-Periodically, LN needs to release a new version of software and coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per instructions.
+In order to release a new version of software, Libra Network (LN) needs to coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per the provided instructions.
```
Comment (aching) [inline lips/lip-3.md:27]:
```suggestion
In order to release a new version of software, Libra Networks (LN) needs to coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per the provided instructions.
```

**Review by aching — APPROVED** [review]

File: lips/lip-3.md [hunk 10]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
 Due to being a reserve backed currency, the true blockchain is the one that is maintained by the Libra Association and reflects the assets in the Libra Reserve. In other words, in contrast to Bitcoin and other “proof of work” and “proof of stake” networks where the miners and users beliefs determine the current fork of the blockchain, with Libra the blockchain used by the Libra Reserve determines the true blockchain. Furthermore, Libra Networks operates as a regulated payment company under a Swiss license and must be able to enforce its policy and oversight, as well as to respond to critical operational incidents.
 
-The goal of LUM is to provide the Association with means to exercise its authority to reconfigure the blockchain and to effectively remediate any plausible operational incident while maintaining the trust gained through decentralized governance and distributed systems.
+The goal of LUM is to provide the means to effectively and transparently update the software in order to meet changing requirements as well as remediate any plausible operational incident.
+
 We first present all building blocks and the reasoning, followed by concrete examples.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:17]: I would delete this sentence

File: lips/lip-3.md [hunk 11]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
 # Building blocks
-There’s two types of upgrades in the system - (i) updates to the Libra protocol software and (ii) changes to on-chain state. We discuss them correspondingly.
+There are two types of upgrades in the system - (i) [updates to the Libra protocol software](#software-upgrade) and (ii) [changes to on-chain state](#on-chain-state-upgrade). We discuss them correspondingly.
+
+All building blocks may be used for "business-as-usual" upgrades or to address urgent incidents.
 
 ## Software upgrade
 Specification changes go through the LIP process and non-specification changes (e.g. bug fixes and performance improvements) do not need to initiate the LIP process. All software changes are submitted through standard Github reviews and proceed through the code review process before being merged to master.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:25]: 1. pls add ref to LIP-0 
2. initiate the LIP process --> go through it

File: lips/lip-3.md [hunk 12]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
-Periodically, LN needs to release a new version of software and coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per instructions.
+In order to release a new version of software, Libra Networks (LN) needs to coordinate the release plan and timeline with all the validator operators. Validator operators are responsible for deploying the software release as per the provided instructions.
 
-Some specification changes may require all validators to transition at the same height, which is accompanied with an on-chain state value update (described in the next section).
+Some changes may require all validators to transition at the same height, which is accompanied with an [on-chain state update](#on-chain-state-upgrade).
```
Comment (dahliamalkhi) [inline lips/lip-3.md:29]: height -> chain height (or blockchain height)
accompanied with --> implemented by

File: lips/lip-3.md [hunk 13]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
 ## On-chain state upgrade
-On-chain state can only be updated via transactions. We propose three ways for LN to execute critical updates, ordered by most frequently expected to least frequently expected: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
+On-chain state can only be updated via transactions. We propose three ways to execute critical updates, ordered by most frequent to least frequent: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:32]: why "critical updates"? these could either be "business as usual" or urgent. How about:

.....three ways [to effectuate changes to the LPN by directly mutating on-chain state].

File: lips/lip-3.md [hunk 14]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
-The goal of LUM is to provide the Association with means to exercise its authority to reconfigure the blockchain and to effectively remediate any plausible operational incident while maintaining the trust gained through decentralized governance and distributed systems.
+The goal of LUM is to provide the means to effectively and transparently update the software in order to meet changing requirements as well as remediate any plausible operational incident.
+
 We first present all building blocks and the reasoning, followed by concrete examples.
 
 # Building blocks
-There’s two types of upgrades in the system - (i) updates to the Libra protocol software and (ii) changes to on-chain state. We discuss them correspondingly.
+There are two types of upgrades in the system - (i) [updates to the Libra protocol software](#software-upgrade) and (ii) [changes to on-chain state](#on-chain-state-upgrade). We discuss them correspondingly.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:20]: I propose the following start:


Conceptually, the software running on validator nodes comprises two layers. One is a *blockchain core* whose goal is to replicate a ledger of transactions and their outcome. The second is a *framework* that stores the ledger-state and a set of software modules  -- encoded in the Move programming language -- that define the rules for interacting with the LPN and mutating the ledge-state (see also LIP-2[ref]). There are two corresponding type of upgrades to the system, (i) updates to the core software (ii) updates to the framework state

File: lips/lip-3.md [hunk 15]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
-On-chain state can only be updated via transactions. We propose three ways for LN to execute critical updates, ordered by most frequently expected to least frequently expected: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
+On-chain state can only be updated via transactions. We propose three ways to execute critical updates, ordered by most frequent to least frequent: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
 
 ### Normal multi-sig transactions
-Normal transactions go through VM and smart contracts, both the resulting state change and the authenticating set of signatures are transparent and visible on-chain for auditing.
-
-Different groups would authenticate different types of transactions. For example, there would be a group for compliance transactions and another group for validator set management.
+Normal transactions go through the VM and invoke Libra Framework modules (see [LIP-2](./lip-2.md)) to produce a Writeset that alters ledger state. Both the resulting state change and the authenticating set of signatures are transparent and visible on-chain for auditing.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:35]: alters <the> ledger-state.

File: lips/lip-3.md [hunk 16]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
 ### Multi-sig WriteSet transactions
-Writeset transactions are a special type of multi-sig transactions that prescribes an arbitrary state change. Unlike normal transactions, which must follow rules such as “the reserve is the only entity that can create or destroy money”, WriteSet transactions simply describe a state change rather than generate one using the blockchain logic that is encoded in Move. This type of transaction is only needed for extremely rare situations such as updating immutable stdlib libraries due to critical bugs or reversing large scale fraud.
+Writeset transactions are a special type of multi-sig transactions that prescribes an arbitrary state change. Unlike normal transactions, which must follow rules such as “the reserve is the only entity that can create or destroy money”, WriteSet transactions simply describe a state change rather than generate one using the blockchain logic that is encoded in Move. This type of transaction is only needed for situations such as updating immutable stdlib libraries due to critical bugs or reversing large scale fraud.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:38]: "than generate one using the blockchain logic that is encoded in Move" --> than go through the Framework modules/rules.

File: lips/lip-3.md [hunk 17]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
 The use of WriteSet transactions follows the spirit of transparency — WriteSet transactions allow an arbitrary change to the state of the blockchain without altering the history of past transactions.
 Even though WriteSet transactions can describe arbitrary state change, they are authenticated with the same multi-sig mechanism as normal transactions which means both the change and multi-sig are transparent and visible on-chain.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:41]: can describe --> prescribe

File: lips/lip-3.md [hunk 18]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
+Genesis transactions can be either normal or WriteSet transactions. As the name suggests, the process is similar to how the very first genesis transaction works. The genesis transaction applies on top of a specific snapshot of a Libra blockchain (might be empty). LN is responsible for picking the appropriate snapshot and building the genesis transactions. Users of Libra including validators use the genesis transaction to authenticate the blockchain. If a validator fails to acquire the correct genesis transaction, it will vote on an irrelevant blockchain, but not otherwise impact the system. If a user of the system acquires the incorrect genesis transaction, they will see incorrect state.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:44]: "the process.." of generating a genesis transaction?

File: lips/lip-3.md [hunk 19]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
 Note after Libra starts, we may still need genesis transactions as the last resort for extremely unlikely catastrophic scenarios such as losing more than 1/3 validators.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:46]: Note -> Note that, (comma)
omit "still"
the last resort -> a last resort
1/3 -> a third

File: lips/lip-3.md [hunk 20]
```diff
@@ -8,52 +8,51 @@ created: 06/09/2020
...
 
-# Case study
-
-Here's a decision flow chart that summarizes the building blocks and how they will be used in specific examples.
+The decision flow chart below summarizes the building blocks and how they will be used in specific examples.
```
Comment (dahliamalkhi) [inline lips/lip-3.md:48]: omit "in specific examples"

**Review by dahliamalkhi — COMMENTED** [review]
@zekun000 please see stylistic comments

**Pull request merged**